### ISSUE TITLE
Will Deep Implicit Models ever become first class citizens in PyTorch?

### ISSUE THREAD

**Issue by randolf-scholz** [pytorch/pytorch#100626]
### 🚀 The feature, motivation and pitch

PyTorch makes it unnecessarily difficult to work with [Deep Implicit Models](https://implicit-layers-tutorial.org/). Deep Implicit Layers/Models are characterized by the need to implement custom backward passes to be efficient, making use of the [implicit function theorem](https://en.wikipedia.org/wiki/Implicit_function_theorem).

In particular, the following issues persist:

- JIT is incompatible with custom backwards. The only way around this is writing custom C++ extensions.  https://github.com/pytorch/pytorch/issues/35749
- Implementing custom backward hooks for a whole `nn.Module` is extremely counter-intuitive, consider this example from Chapter 4:

```python
def forward(self, x):
    # compute forward pass and re-engage autograd tape
    with torch.no_grad():
        z, self.forward_res = self.solver(
            lambda z : self.f(z, x), torch.zeros_like(x), **self.kwargs)
    z = self.f(z,x)
    
    # set up Jacobian vector product (without additional forward calls)
    z0 = z.clone().detach().requires_grad_()
    f0 = self.f(z0,x)
    def backward_hook(grad):
        g, self.backward_res = self.solver(
            lambda y : autograd.grad(f0, z0, y, retain_graph=True)[0] + grad,
            grad, **self.kwargs)
        return g
            
    z.register_hook(backward_hook)
    return z
```

One needs to `no_grad` the forward, and then manually surgically insert the custom backward.

- Custom backward of nn.Modules are also incompatible with JIT, and it's unclear how to work around it.

### Alternatives

JAX supports per-function custom backward overrides via decorators: https://jax.readthedocs.io/en/latest/notebooks/Custom_derivative_rules_for_Python_code.html

```python
import jax.numpy as jnp
from jax import custom_jvp

@custom_jvp
def f(x, y):
  return jnp.sin(x) * y

@f.defjvp
def f_jvp(primals, tangents):
  x, y = primals
  x_dot, y_dot = tangents
  primal_out = f(x, y)
  tangent_out = jnp.cos(x) * x_dot * y + jnp.sin(x) * y_dot
  return primal_out, tangent_out
```

If supported, this would solve all 3 issues.

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @Lezcano @Varal7 @soumith @msaroufim @wconstab @ngimel @bdhirsh @Chillee @samdow @kshitij12345 @janeyx99

**Comment by vadimkantorov** [comment]
A little bit related:
- https://github.com/pytorch/pytorch/issues/50688
- https://github.com/pytorch/pytorch/issues/23756

**Comment by ngimel** [comment]
cc @zou3519, seems related to custom ops.
Aside from not being supported by jit, does custom AutogradFunction do what you want?

**Comment by albanD** [comment]
> Implementing custom backward hooks for a whole nn.Module is extremely counter-intuitive, consider this example

I'm not sure to follow this part? Why just wrapping the whole forward in a custom Function enough?

> JIT is incompatible with custom backwards. 

As Natalia mentioned, this won't be fixed, not using jit will avoid the problem!

**Comment by randolf-scholz** [comment]
@albanD @kshitij12345 @ngimel 

I don't think it is possible to implement a DEQ layer as a `autograd.Function` currently.  Recall that for a Deep Equilibrium Layer, we have some neural network $f$ and we compute:

1. **Forward:** given input $x$, return solution  $z⁎$  of the Fixed Point equation $z = f(z，x，θ)$.
2. **Backward:** Given outer gradients $y$, we need  to compute the VJP $\frac{∂z⁎}{∂θ}^⊤y$. This is done as follows, in order to avoid backpropagating through the solver in the forward pass:
   1. Compute solution $g⁎$ of the linear system $g+\text{VJP}(f, z⁎，g) = y$.
   2. Compute $\text{VJP(z⁎，θ，y)} = \text{VJP}(f，θ，g⁎)$.

Again, check out chapter 4 of https://implicit-layers-tutorial.org/ for a derivation.

So, what we want is a `autograd.Function` that takes two inputs: 

1. A `nn.Module` $f$
2. A `Tensor` $x$ 

and returns the computed fixed point. I looked but couldn't find a single example where people input a whole network to `autograd.Function.forward`; it doesn't seem to be intended. The documentation says:

> All kinds of Python objects are accepted here. Tensor arguments that track history (i.e., with requires_grad=True) will be converted to ones that don’t track history before the call, and their use will be registered in the graph. **Note that this logic won’t traverse lists/dicts/any other data structures and will only consider tensors that are direct arguments to the call.**

Which seems to be a showstopper to using `autograd.Function` for this purpose.

I created a demo notebook with my failed attempt to use `autograd.Function` for this purpose: [deq_autograd_function.zip](https://github.com/pytorch/pytorch/files/11431355/deq_autograd_function.zip). However, while the manual gradient computation works, wrapping everything in a `autograd.Function` only results in `None` outputs. I am pretty sure if it were possible to do it with `autograd.Function`, people wouldn't be using this counter-intuitive hack using the `register_hook`.

<details><summary>Attempted implementation via `autograd.Function`</summary>

```python
#!/usr/bin/env python
# coding: utf-8
# %% [markdown]
# # Test of the possibility of inputting a nn.Module as input into a custom autograd function

# %%
import warnings
from typing import Any, Callable, NamedTuple, Optional

import numpy as np
import torch
from torch import Tensor, dot, nn
from torch.autograd import grad


# %% [markdown]
# # Setup - Implement a Linear Solver (CGS - Conjugate Gradients Squared)

# %%
class CGS_STATE(NamedTuple):
    """State of the conjugate gradient squared solver."""

    L: Callable[[Tensor], Tensor]
    """The linear function."""
    x: Tensor
    """Vector: Current iterate."""
    r: Tensor
    """Vector: Residual vector."""
    p: Tensor
    """Vector: Search direction."""
    q: Tensor
    """Vector: """
    r0star: Tensor
    """Vector: Initial dual residual vector."""
    rho: Tensor
    """Scalar: Inner Product between r and r0star."""


def cgs_step(state: CGS_STATE) -> CGS_STATE:
    """Perform a single step of the conjugate gradient squared method."""
    # unpack state
    L = state.L
    x = state.x
    r = state.r
    p = state.p
    q = state.q
    r0star = state.r0star
    rho_old = state.rho

    # perform iteration
    rho = dot(r, r0star)
    beta = rho / rho_old
    u = r + beta * q
    p = u + beta * (q + beta * p)
    v = L(p)
    sigma = dot(v, r0star)
    alpha = rho / sigma
    q = u - alpha * v
    r = r - alpha * L(u + q)
    x = x + alpha * (u + q)
    return CGS_STATE(L=L, x=x, r=r, p=p, q=q, r0star=r0star, rho=rho)


def cgs(
    L: nn.Module,
    y: Tensor,
    x0: Optional[Tensor] = None,
    r0star: Optional[Tensor] = None,
    maxiter: int = 100,
    atol: float = 10**-8,
    rtol: float = 10**-5,
) -> CGS_STATE:
    """Solves linear equation L(x)=y."""
    tol = max(atol, rtol * y.norm())
    x0 = torch.zeros_like(y) if x0 is None else x0
    r0 = y - L(x0)
    r0star = r0.clone() if r0star is None else r0star
    p0 = torch.zeros_like(r0)
    q0 = torch.zeros_like(r0)
    rho0 = 1.0  # dot(r0, r0star)
    state = CGS_STATE(L=L, x=x0, r=r0, p=p0, q=q0, r0star=r0star, rho=rho0)

    for it in range(maxiter):
        state = cgs_step(state)

        if state.r.norm() <= tol:
            print(f"Converged after {it} iterations.")
            break
    else:
        warnings.warn(f"No convergence after {maxiter} iterations.")

    residual = (y - L(state.x)).norm().item()
    print(f"Final {residual=:.4}  (r={state.r.norm().item():.4})")
    return state


# %% [markdown]
# ## Test whether CGS works

# %%
N = 8
L = nn.Linear(N, N, bias=False)
# L.weight = nn.Parameter(torch.eye(N) + torch.randn(N, N) / np.sqrt(N))
y = torch.randn(N)
x0 = torch.zeros_like(y)
x_cgs = cgs(L, y).x

# %% [markdown]
# ## Verify against scipy

# %%
from scipy.sparse.linalg import cgs as cgs_scipy

A = L.weight.detach().numpy()
b = y.numpy()
x_ref, r = cgs_scipy(A, b, x0=np.zeros_like(b))
print(f"Final residual: {np.linalg.norm(A @ x_ref - b)}")

diff = np.mean((x_cgs.detach().numpy() - x_ref) ** 2)
print(f"MSE between custom and reference solution: {diff}")


# %% [markdown]
# # Test on a model
#
# We compute gradients for $‖\text{deq-layer}(x)‖^2$.

# %%
def fixed_point_iteration(
    f: nn.Module,
    x: Tensor,
    maxiter: int = 100,
    atol: float = 10**-8,
    rtol: float = 10**-5,
) -> Tensor:
    """Solves $z⁎=f(x，z⁎)$ via FP iteration."""
    z = torch.zeros(f.hidden_size)
    for it in range(maxiter):
        z_new = model(x, z)
        converged = (z_new - z).norm() <= rtol * z.norm() + atol
        z = z_new
        if converged:
            print(f"Converged after {it} iterations.")
            break
    else:
        warnings.warn(f"No convergence after {maxiter} iterations.")
    return z


input_size, hidden_size = 4, 3
model = nn.RNNCell(input_size=input_size, hidden_size=hidden_size)
x = torch.randn(input_size)

# %% [markdown]
# ## 1. Automatic differentiation (through the fixed-point iteration)

# %%
model.zero_grad(set_to_none=True)

# forward: fixed point iteration
z = fixed_point_iteration(model, x)

# backward
z.norm().pow(2).backward()
reference_gradients = [w.grad for w in model.parameters()]
print(reference_gradients)

# %% [markdown]
# ## 2. Manual computation

# %%
model.zero_grad(set_to_none=True)

# forward: fixed point iteration
with torch.no_grad():
    z = fixed_point_iteration(model, x)

# backward setup.
outer_grad = 2 * z  # ∂‖z‖²/∂z = 2z
zstar = z.requires_grad_()  # must enable grad to compute ∂f/∂z⁎
fstar = model(x, zstar)
params = list(model.parameters())

# backward step 1: solve for $g ≔ \Bigl(𝕀 - \frac{∂f}{∂z⁎}\Bigr)^{-⊤} y$
L = lambda g: g - grad(fstar, zstar, g, retain_graph=True)[0]
gstar = cgs(L, outer_grad).x

# compute the outer grad
manual_gradients = grad(fstar, params, gstar)

print("MSE between automatic gradients to manual gradients:")
for g1, g2 in zip(reference_gradients, manual_gradients):
    print((g1 - g2).pow(2).mean())


# %% [markdown]
# ## 3. Using a custom autograd function

# %%
class DEQ_Layer(torch.autograd.Function):
    @staticmethod
    def forward(f: nn.Module, x: Tensor, **kwargs: Any) -> Tensor:
        """Compute the fixed point $z⁎ = f(x, z⁎)$."""
        zstar = fixed_point_iteration(f, x, **kwargs)
        return zstar.requires_grad_()

    @staticmethod
    def setup_context(ctx, inputs, output, **_):
        f, x = inputs
        zstar = output
        ctx.save_for_backward(x, zstar)
        ctx.f = f
        # fstar = f(x, zstar)
        # params = list(f.parameters())
        # ctx.save_for_backward(zstar, fstar, params)

    @staticmethod
    def backward(self, ctx, grad_output):
        """Solve the linear system $(𝕀 - ∂f(x，z⁎)/∂z⁎)ᵀg = y$."""
        x, zstar = ctx.saved_tensors
        f = ctx.f

        with torch.enable_grad():
            fstar = f(x, zstar)

        # solve the linear system
        L = lambda g: g - grad(fstar, zstar, g, retain_graph=True)[0]
        gstar = cgs(L, grad_output).x
        return None
        # compute the outer grad
        return grad(fstar, params, gstar)


# %%
model.zero_grad(set_to_none=True)

# forward
y = DEQ_Layer.apply(model, x).norm().pow(2)

# backward
y.backward()
print([w.grad for w in model.parameters()])  # ✘ no gradients...
model.zero_grad()

# %% [markdown]
# ## Background: DEQ models
#
# In a Deep Equilibrium model, given input $x$, we use out model $f$, parameterized by $θ$ to compute the fixed point
#
#
# $$ z⁎ = f(z⁎，x，θ) $$
#
# Now, we need the gradients $\frac{∂z⁎}{∂θ}$. Computing the derivative on both sides yields
#
# $$ \frac{∂z⁎}{∂θ} 
# = \frac{∂f}{∂z⁎}\frac{∂z⁎}{∂θ} + \frac{∂f}{∂x}\frac{∂x}{∂θ} + \frac{∂f}{∂θ}\frac{∂θ}{∂θ}
# = \frac{∂f}{∂z⁎}\frac{∂z⁎}{∂θ} + \frac{∂f}{∂θ}
# ⟹ \Bigl(𝕀 - \frac{∂f}{∂z⁎}\Bigr)\frac{∂z⁎}{∂θ} = \frac{∂f}{∂θ}
# $$
#
# In particular, the VJP given outer gradient $g$ is
#
# $$ \frac{∂z⁎}{∂θ}^⊤ y = \frac{∂f}{∂θ}^⊤ \Bigl(𝕀 - \frac{∂f}{∂z⁎}\Bigr)^{-⊤} y $$
#
# so, as an intermediate we need to compute 
#
# $$ g ≔ \Bigl(𝕀 - \frac{∂f}{∂z⁎}\Bigr)^{-⊤} y ⟺ \Bigl(𝕀 - \frac{∂f}{∂z⁎}\Bigr)^⊤ g = y ⟺ g + \text{VJP}(f, z⁎, g) = y$$
#
# Once we have $g$ we can compute 
#
# $$\text{VJP(z⁎，θ，y)} = \text{VJP}(f，θ，g)$$
#
# In summary, the steps for computing the gradients are:
#
# 1. **Forward:** given input $x$, return solution  $z⁎$  of the Fixed Point equation $z = f(z，x，θ)$.
# 2. **Backward:** Given outer gradients y, we need  to compute the gradients $\frac{∂z⁎}{∂θ}$.
#    1. Compute solution $g⁎$ of the linear system $g+\text{VJP}(f, z⁎，g) = y$.
#    2. Compute $\text{VJP(z⁎，θ，y)} = \text{VJP}(f，θ，g⁎)$.In summary, the steps for computing the gradients are:
```

</details>

**Comment by albanD** [comment]
Hey!
Thanks for the pointers. After reading the paper, I would agree with what the author states there: "This means that implementing a generic DEQ layer, capable of using any function f, inside a library like PyTorch is actually very straightforward."

You can actually compute a rather complex quantity without having to re-implement any backward yourself. Also the implementation that you shared above will work with any function `f` so it shouldn't be a problem that it has to do some non-trivial logic?
In general, this logic will need to live somewhere between the solver code and the user API. This kind of simple Module's forward that can wrap any solve and function sounds like a good abstraction to me.

Could you give details on what you think is limiting with using such abstraction?

**Comment by randolf-scholz** [comment]
> I'm not sure what is the concern here. PyTorch is built such that you can do these during the forward and don't have to worry about it!

Well, I usually like to JIT the code because (1) speed (which can be quite a bit for RNN-type models) and (2) serializability (JIT serialization is much nicer than `state_dict`, because it only requires the model checkpoint and torch, but not the original code)

> I'm not sure how to get around this. You do want to compute a special quantity that is not a gradient or anything. So it is expected that you'll have to use tricks to compute it. In this case, you're leveraging pytorch's autograd to get very specific quantities as needed. So it is expected that you will need to do surgery to compute these specific quantities.

Oh, we do want to compute a gradient, this is what this is all about! 

We want to calculate $\frac{∂𝓛}{∂θ} = \frac{∂𝓛}{∂z⁎}\frac{∂z⁎}{∂θ}$, where $z⁎$ is the solution to the equation $z=f(z, x, θ)$. Using the implicit function theorem, we can derive that $ \text{VJP}(z⁎，θ，y)$ is equivalent to $\text{VJP}(f，θ，g⁎)$ were $g⁎$ is the solution of the linear equation $g+\text{VJP}(f, z⁎，g) = y$, and what we want to do is implement the backward using this formula, instead of letting autograd backprop through the forward solve. (Neural ODEs are another example of this technique). This is exactly what is shown e.g. in section 2 of the notebook I attached. 

Now the "surgery" approach is using the `register_hook` in the forward pass. What would be more intuitive, imo, is to allow adding a completely custom `backward` to a whole `nn.Module`.

(And as a side note: sometimes one would want to add custom backwards to other methods than the forward, for example the inverse pass of an invertible ResNet.)

**Comment by albanD** [comment]
> Now the "surgery" approach is using the register_hook in the forward pass. What would be more intuitive, imo, is to allow adding a completely custom backward to a whole nn.Module.

As said above, you don't have to use hooks. It's just a nice trick to avoid having to define a custom backward by hand.

As said above as well, if you do want to write a backward by hand for any function (including Module methods), you should use autograd.Function at that function level (so it takes Tensor as inputs, not Module!). In any case, this is just a code organization detail and I do not know of any limitation coming from this.

> Well, I usually like to JIT the code

I'm afraid the scope of this discussion is way beyond this issue. I would refer you to this post that discusses this in details: https://dev-discuss.pytorch.org/t/next-steps-for-pytorch-compilers/309

**Comment by randolf-scholz** [comment]
> As said above, you don't have to use hooks. It's just a nice trick to avoid having to define a custom backward by hand.

I'd love to define the backward by hand in this specific case, the hook thing is counter-intuitive.

> use autograd.Function at that function level (so it takes Tensor as inputs, not Module!)

This is undesirable, as it would mean one would effectively have to reimplement the whole forward-code of $f$ inside the DEQ layer, breaking modularity. But I feel like we are talking in circles now, so probably it's just better to close this at this point and either deal with the counter-intuitiveness and JIT-limitations or reimplement in JAX.

**Comment by Chillee** [comment]
@albanD I think it's not a bad idea to add an alternative wrapper to autograd.Function that allows users to pass in modules/other wrapper types. 

@randolf-scholz It's also possible to write it yourself. I just hacked up how I would do it. It's a bit finicky since you need to manually handle flattening/unflattening of the parameters yourself, which is why I think it's plausible for us to add a wrapper here.

```
import torch
import torch.nn as nn
import torch.utils._pytree as pytree

class MLP(torch.nn.Module):
    def __init__(self, hidden_size):
        super().__init__()
        self.fc1 = torch.nn.Linear(hidden_size, hidden_size)
        self.fc2 = torch.nn.Linear(hidden_size, hidden_size)

    def forward(self, x):
        x = torch.nn.functional.relu(self.fc1(x))
        x = self.fc2(x)
        return x

def solver(f, x):
    for _ in range(5):
        x = f(x)
    return x

class DEQFixedPoint(nn.Module):
    def __init__(self, f, solver):
        super().__init__()
        self.f = f
        self.solver = solver

        _, param_spec = pytree.tree_flatten({**dict(f.named_parameters()), **dict(f.named_buffers())})

        class DEQFixedPointForward(torch.autograd.Function):
            @staticmethod
            def forward(x, *params_and_buffers):
                def functional_call(x):
                    return torch.func.functional_call(self.f, pytree.tree_unflatten(params_and_buffers, param_spec), x)
                with torch.no_grad():
                    z = solver(functional_call,  x)
                return z

            @staticmethod
            def setup_context(ctx, inputs, output):
                ctx.save_for_backward(*inputs, output[0])

            @staticmethod
            def backward(ctx, outer_grad):
                # retrieve the inputs and outputs from forward
                params_and_buffers, z = ctx.saved_tensors
                # Unimplemented since I'm lazy, but I think you have everything you need
                return x

        self.forward_func = DEQFixedPointForward.apply

    def forward(self, x):
        flat_params_and_buffers = pytree.tree_flatten({**dict(self.f.named_parameters()), **dict(self.f.named_buffers())})[0]
        return self.forward_func(x, *flat_params_and_buffers)

DEQFixedPoint(MLP(10), solver)(torch.randn(10))
```

I'm gonna reopen this issue so we discuss more about adding some kind of wrapper like this to make it easier to deal with modules in autograd.Function.

**Comment by randolf-scholz** [comment]
@Chillee I integrated your approach into my example notebook: [deq_autograd_function.zip](https://github.com/pytorch/pytorch/files/11441896/deq_autograd_function.zip), thanks for making me aware of the `functional_call` method.

<details> <summary>I implemented it with a layer factory pattern:</summary>

```python
from functools import partial, wraps

import torch.utils._pytree as pytree


def rpartial(func, /, *fixed_args, **fixed_kwargs):
    r"""Fix positional arguments from the right."""

    @wraps(func)
    def _wrapper(*func_args, **func_kwargs):
        return func(*(func_args + fixed_args), **(func_kwargs | fixed_kwargs))

    return _wrapper


def deq_layer_factory(
    module: nn.Module,
    fsolver=fixed_point_iteration,
    fsolver_kwargs={},
    bsolver=cgs,
    bsolver_kwargs={},
) -> Callable[[Tensor], Tensor]:
    """Create functional deq_layer for the given module."""

    params, param_spec = pytree.tree_flatten(
        {**dict(module.named_parameters()), **dict(module.named_buffers())}
    )

    fsolver_kwargs = fsolver_kwargs | {"z0": torch.zeros(module.hidden_size)}

    def func(x: Tensor, z: Tensor, *params_and_buffers) -> Tensor:
        """Function call $f(x, z, θ)$."""
        theta = pytree.tree_unflatten(params_and_buffers, param_spec)
        return torch.func.functional_call(module, theta, (x, z))

    class DEQ_Layer(torch.autograd.Function):
        @staticmethod
        def forward(x: Tensor, *params_and_buffers) -> Tensor:
            """Compute the fixed point $z⁎ = f(x, z⁎)$."""
            f = rpartial(func, *params_and_buffers)
            return fsolver(f, x, **fsolver_kwargs)

        @staticmethod
        def setup_context(ctx, inputs, output):
            x, *params_and_buffers = inputs
            zstar = output
            f = rpartial(func, *params_and_buffers)

            with torch.enable_grad():
                # NOTE: without detach, we get an infinite loop.
                zstar = zstar.detach().requires_grad_()
                fstar = f(x, zstar)

            ctx.save_for_backward(fstar, zstar, x)

        @staticmethod
        def backward(ctx, grad_output):
            fstar, zstar, x = ctx.saved_tensors

            # solve the linear system $(𝕀 - ∂f(x，z⁎)/∂z⁎)ᵀg = y$
            L = lambda g: g - grad(fstar, zstar, g, retain_graph=True)[0]
            gstar = bsolver(L, grad_output, **bsolver_kwargs).x

            # compute the outer grads
            grad_x = (
                grad(fstar, x, gstar, retain_graph=True) if x.requires_grad else None
            )
            grad_f = grad(fstar, params, gstar)
            return grad_x, *grad_f

    return rpartial(DEQ_Layer.apply, *params)
```

</details>

It really shows well that the point is that a `nn.Module` really can be considered as a partial function $f_θ(x) ≔ f(x, θ)$. This suggests that one should be able to define `backward` for an `nn.Module`, which must return the gradients for `self` and all inputs of `forward`. When constructing the graph, the autograd engine could check if `backward` is defined for the module. If it is, use it as a plug-in replacement for the automatic backward. 

<details><summary> This would be the cleanest solution imo: </summary>

```python
class DEQFixedPoint(nn.Module):
    def __init__(self, f, solver):
        super().__init__()
        self.f = f
        self.solver = solver

    def forward(self, x: Tensor) -> Tensor:
        zstar =  fixed_point_iteration(self.f, x)
        return zstar

    def backward(self, outer_grad, inputs, output):
        """Computes gradients for inputs of forward.

        Since this a nn.Module, we ought to compute gradients both for 'self' and 'x'
        """
        x = inputs
        zstar = output

        # compute the gradients
        grad_self = ...
        grad_x = ...
        return grad_self, grad_x
```

</details>

<details> <summary>because it is closest to the manual backward.</summary>

```python
# manual backward
model.zero_grad(set_to_none=True)

# forward: fixed point iteration
with torch.no_grad():
    z = fixed_point_iteration(model, x)

# backward setup.
outer_grad = 2 * z  # ∂‖z‖²/∂z = 2z
zstar = z.requires_grad_()  # must enable grad to compute ∂f/∂z⁎
fstar = model(x, zstar)

# backward step 1: solve for $g ≔ \Bigl(𝕀 - \frac{∂f}{∂z⁎}\Bigr)^{-⊤} y$
L = lambda g: g - grad(fstar, zstar, g, retain_graph=True)[0]
gstar = cgs(L, outer_grad).x

# compute the outer grad
manual_gradients = grad(fstar, model.parameters(), gstar)
```

</details>

**Comment by cgoemaere** [comment]
@randolf-scholz I found a way to write up a DEQ elegantly in PyTorch. My code is available [here](https://github.com/cgoemaere/hamdeq/tree/master/hamdeq/deq_core), as part of the codebase for my paper.

I had painstakingly put together the code using functional_call, when I found a way to do it cleaner, using Modules, as you wanted. The main trick is to define the autograd Function within the init of a different class definition, very similar to what was previously suggested. Since Function uses only static methods, you can even access the outer class object using self.

Some code to explain (this is basically [this file](https://github.com/cgoemaere/hamdeq/blob/master/hamdeq/deq_core/deq.py)):
```python
class DEQ(torch.nn.Module):
    def __init__(self, *args):
        super().__init__()

        # Define whatever Module you'd like
       self.deq_module = ...

        class AutogradDEQ(torch.autograd.Function):
            @staticmethod
            def forward(x):
                z_star = solve(self.deq_module, x) # Note that self is the outer object belonging to class DEQ
                return z_star

            @staticmethod
            def setup_context(ctx, inputs, z_star):
                (x,) = inputs
                ctx.save_for_backward(x, z_star)

            @staticmethod
            @torch.autograd.function.once_differentiable
            def backward(ctx, grad_output):
                # Retrieve stored tensors
                x, z_star = ctx.saved_tensors

                # The grads of self.deq_module need to be populated within calc_grads
                x_grad = calc_grads(grad_output, self.deq_module, x, z_star)
                return (x_grad,)

        # Set forward method to be our autograd module
        self.forward: Callable = AutogradDEQ.apply  
```

Also, very recently, the original DEQ research team released [torchdeq](https://github.com/locuslab/torchdeq), which will probably become the go-to repo for DEQs in PyTorch.

Hope this helps!
